### DISCUSSION TITLE
Add linked-identities to id_token

### DISCUSSION THREAD

**Discussion by robrez** [keycloak/keycloak#29992] [Q&A] [answered]
I'm seeking some guidance on how I can add a claim to the id_token (and/or userinfo endpoint) which contains the data from `/linked-accounts`.  Although the modeling from `linked-accounts`, wouldn't quite be what I wanted because the "providerName" is not appealing, I'd prefer "iss".  The gist of what I'd want would be an array.

```json
{
  "sub": "51cf15b0-4381-4fb4-a4b1-592750529dd0",
  "upn": "rob",
  "email_verified": true,
  "linked": [
    "sub@iss",
    "rob@https://token.actions.githubusercontent.com"
  ]
}
```

Other similar flavors, might include making that an array of objects rather than strings. Or creating dedicated claims for a specific provider.. eg:

```json
{
  "sub": "51cf15b0-4381-4fb4-a4b1-592750529dd0",
  "upn": "rob",
  "email_verified": true,
  "urn:foo:linked-account:github": "rob"
}
```

**Comment by danielFesenmeyer** [comment, marked as answer]
It seems like Keycloak does not have builtin functionality for your use case.

You could implement your own "LinkedAccountsMapper", similar to builtin mappers such as UserAttributeMapper or GroupMembershipMapper.
You can get the "linked identities" like this (code copied from UserResource):
```
    private Stream<FederatedIdentityRepresentation> getFederatedIdentities(UserModel user) {
        Set<String> idps = realm.getIdentityProvidersStream().map(IdentityProviderModel::getAlias).collect(Collectors.toSet());
        return session.users().getFederatedIdentitiesStream(realm, user)
                .filter(identity -> idps.contains(identity.getIdentityProvider()))
                .map(ModelToRepresentation::toRepresentation);
    }
```

**Comment by robrez** [comment]
> Or creating dedicated claims for a specific provider..

```json
{
  "sub": "51cf15b0-4381-4fb4-a4b1-592750529dd0",
  "upn": "rob",
  "email_verified": true,
  "urn:foo:linked-account:github": "rob"
}
```

I did find a built-in way to achieve this flavor.  It can be done by adding an "Attribute Importer" Mapper to the identity provider. However, I would still prefer something more similar to the array-based solution

**Reply by danielFesenmeyer** [reply]
Seems to be a nice approach, in case you don't have many IDPs.
